impl(otel): copy service resource labels into metric labels

IcySteam opened the pull request:
fixed #14823

<!-- Reviewable:start -->
- - -
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/googleapis/google-cloud-cpp/14825)
<!-- Reviewable:end -->

dbolduc reviewed (changes requested):
I don't think the copying is necessary. Can we check for the known service labels and do the manipulations here:

https://github.com/googleapis/google-cloud-cpp/blob/912533cf46cc0e9e08c90e12472a65d9adaffbc9/google/cloud/opentelemetry/internal/time_series.cc#L174-L177

and add a `TEST(ToMonitoredResource, AddsServiceLabels) { ... }` in `time_series_test.cc` ?

dbolduc commented on google/cloud/opentelemetry/internal/monitored_resource.cc:
```diff
@@ -229,6 +233,52 @@ MonitoredResource ToMonitoredResource(
+         scope_metrics.metric_data_) {
+      for (opentelemetry::sdk::metrics::PointDataAttributes& pda :
+           metric.point_data_attr_) {
+        auto& attributes = pda.attributes;
```
Sorry, I am confused. Do we want the labels to end up getting added to the `google.api.Metric` https://github.com/googleapis/googleapis/blob/2d05911be5a468b556236bee537c91922f9c23a3/google/api/metric.proto#L286

... or to the `google.api.MonitoredResource` https://github.com/googleapis/googleapis/blob/2d05911be5a468b556236bee537c91922f9c23a3/google/api/monitored_resource.proto#L106

... or both?

unit tests would help clarify the intended behavior that we want.

dbolduc commented:
/gcbrun

IcySteam commented:
Thanks; we indeed want the labels to end up getting added to the `google.api.Metric` class. I've refactored the PR and added some unit tests in the latest commit, PTAL.

dbolduc commented on google/cloud/opentelemetry/internal/time_series.h:
```diff
@@ -34,6 +38,15 @@ GOOGLE_CLOUD_CPP_INLINE_NAMESPACE_BEGIN
 // See: https://cloud.google.com/monitoring/quotas
 auto constexpr kMaxTimeSeriesPerRequest = 200;
 
+std::unordered_map<std::string, OTelKeyMatch> const kExtraLabelsLookup = {
```
On second thought, I don't think this struct is buying us much of anything. It seems kind of misapplied, in that the things like `"service_name"` are supposed to represent monitored resources.

I think we will be better off just iterating over the 3 labels we care about, and reuse the name changing logic from `ToMetric()`

dbolduc commented on google/cloud/opentelemetry/internal/time_series_test.cc:
```diff
@@ -697,6 +711,82 @@ TEST(ToRequests, Batches) {
+  rm.scope_metric_data_.push_back(std::move(sm));
+
+  auto tss = ToTimeSeries(rm, PrefixWithWorkload);
+  for (auto const& kv : kExtraLabelsLookup) {
+    auto const& oks = kv.second.otel_keys;
+    if (oks.empty()) {
+      continue;
+    }
+    for (auto const& ts : tss) {
```
We should avoid logical branches and stuff in tests. Let's try to make them as simple as possible.

dbolduc commented on google/cloud/opentelemetry/internal/time_series_test.cc:
```diff
@@ -697,6 +711,82 @@ TEST(ToRequests, Batches) {
+      continue;
+    }
+    for (auto const& ts : tss) {
+      auto const& labels = ts.metric().labels();
+      ASSERT_TRUE(labels.find(kv.first) != labels.end());
+      EXPECT_EQ(labels.at(kv.first), kv.first);
```
We probably want to use GMock's container matchers.

https://google.github.io/googletest/reference/matchers.html

dbolduc commented on google/cloud/opentelemetry/internal/time_series_test.cc:
```diff
@@ -697,6 +711,82 @@ TEST(ToRequests, Batches) {
+    for (auto const& ts : tss) {
+      auto const& labels = ts.metric().labels();
+      ASSERT_TRUE(labels.find(kv.first) != labels.end());
+      if (existing_labels.GetAttributes().find(kv.first) ==
```
`if`s in tests kind of scare me.

dbolduc commented on google/cloud/opentelemetry/internal/time_series.cc:
```diff
@@ -215,6 +217,7 @@ std::vector<google::monitoring::v3::TimeSeries> ToTimeSeries(
       }
     }
   }
+  WithExtraLabels(data, tss);
```
I think it would make sense to build this into `ToMetric(...)` which deals specifically with the `google.api.Metric`. It also has logic for changing strings from `service.name` -> `service_name` which we can reuse.

dbolduc reviewed (commented):
Thanks for the PR. Sorry it took me so long to review it. I have been splitting time between `google-cloud-cpp` and another team.

I feel like I owe you one, so I made the review changes in #14930. Hopefully that can cut out some review iterations and unblock y'all faster.

dbolduc commented:
I am closing this PR in favor of #14930 

Thank you @IcySteam for the hard work on this PR. I copied a lot of it in #14930.
